actions/websocket_actions.test.jsx

```diff
@@ -930,21 +969,199 @@ describe('handlePluginEnabled/handlePluginDisabled', () => {
+            server: {
+                executables: {
+                    'linux-amd64': 'server/dist/plugin-linux-amd64',
+                    'darwin-amd64': 'server/dist/plugin-darwin-amd64',
+                    'windows-amd64': 'server/dist/plugin-windows-amd64.exe',
+                },
+            },
+            webapp: {
+                bundle_path: 'webapp/dist/main.js',
+            },
+        };
+
+        let getWebappPlugins;
+        let loadPluginsIfNecessary;
+
+        beforeEach(async () => {
+            console.log = jest.fn();
+            console.error = jest.fn();
+
+            const {Client4} = require('mattermost-redux/client'); // eslint-disable-line global-require
+            getWebappPlugins = Client4.getWebappPlugins;
+            loadPluginsIfNecessary = jest.requireActual('plugins').loadPluginsIfNecessary;
+
+            document.createElement = jest.fn();
+            document.getElementsByTagName = jest.fn();
+            document.getElementsByTagName.mockReturnValue([{
+                appendChild: jest.fn(),
+            }]);
+        });
+
+        test('should remove unused old plugins when socket reconnects', async () => {
+            const manifest = {
+                ...baseManifest,
+                id: 'com.mattermost.demo-2-plugin',
+            };
+
+            const initialize = jest.fn();
+            window.plugins = {
+                [manifest.id]: {
+                    initialize,
+                },
+            };
+            const mockScript = {};
+            document.createElement.mockReturnValue(mockScript);
+
+            expect(mockScript.onload).toBeUndefined();
+
+            // Enable plugin
+            handlePluginEnabled({data: {manifest}});
+            mockState.plugins.manifests = {[manifest.id]: manifest};
+
+            const dispatchReceivedArg = store.dispatch.mock.calls[0][0];
+            expect(dispatchReceivedArg.type).toBe(ActionTypes.ADD_WEBAPP_PLUGIN);
+            expect(dispatchReceivedArg.data).toBe(manifest);
+
+            expect(document.createElement).toHaveBeenCalledWith('script');
+            expect(document.createElement).toHaveBeenCalledTimes(1);
+
+            // Disable plugin with reload
+            mockState.entities.general.config.PluginsEnabled = 'true';
+            getWebappPlugins.mockImplementation(() => []);
+            await loadPluginsIfNecessary();
+            mockState.plugins.manifests = {};
+
+            const dispatchRemovedArg = store.dispatch.mock.calls[1][0];
+            expect(dispatchRemovedArg.type).toBe(ActionTypes.REMOVE_WEBAPP_PLUGIN);
             expect(dispatchRemovedArg.data).toBe(manifest);
+
+            // Assert disabling is idempotent
+            await loadPluginsIfNecessary();
+
+            expect(store.dispatch).toHaveBeenCalledTimes(3);
+            expect(console.error).toHaveBeenCalledTimes(0);
+        });
+
+        test('should upgrade old plugins when socket reconnects', async () => {
```

Comment:
These tests are great!! @Qovaros do you mind adding one for the following case?

1. Where we add a new plugin and remove an existing plugin inside loadPluginsIfNecessary

Reply:
Sure I just added this test.